postgres/grove/data_source.go

```diff
@@ -0,0 +1,230 @@
+package grove
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+
+	"github.com/buildwithgrove/path/envoy/auth_server/proto"
+	"github.com/jackc/pgx/v5"
+	"github.com/jackc/pgx/v5/pgconn"
+	"github.com/jackc/pgx/v5/pgxpool"
+	"github.com/jackc/pgxlisten"
+	"github.com/pokt-network/poktroll/pkg/polylog"
+
+	grpc_server "github.com/buildwithgrove/path-auth-data-server/grpc"
+	"github.com/buildwithgrove/path-auth-data-server/postgres/grove/sqlc"
+)
+
+// postgresDataSource implements the grpc_server.AuthDataSource interface.
+var _ grpc_server.AuthDataSource = &postgresDataSource{}
+
+type (
+	// postgresDataSource implements the AuthDataSource interface for a Postgres database.
+	// The database driver is generated using SQLC: https://docs.sqlc.dev/en/latest/index.html
+	//
+	// The schema defined in ./postgres/sqlc/grove_schema.sql is compatible with the existing
+	// Grove Portal database schema to allow PATH to stream updates from Grove Portal DB.
+	//
+	// For the current Grove Portal DB schema as defined in the Portal HTTP DB (PHD) repo:
+	// https://github.com/pokt-foundation/portal-http-db/blob/master/postgres-driver/sqlc/schema.sql
+	postgresDataSource struct {
+		driver   *postgresDriver
+		listener *pgxlisten.Listener
+
+		notificationCh chan *Notification
+		updatesCh      chan *proto.AuthDataUpdate
+
+		logger polylog.Logger
+	}
+	// The postgresDriver struct wraps the SQLC generated queries and the pgxpool.Pool.
+	// See: https://docs.sqlc.dev/en/latest/tutorials/getting-started-postgresql.html
+	postgresDriver struct {
+		*sqlc.Queries
+		DB *pgxpool.Pool
+	}
+)
+
+/* ---------- Postgres Connection Funcs ---------- */
```

Comment:
Can you add a TODO_IMPROVE to generalize these for other postgres users in the future?